### PR TITLE
[Mockito] Fix Unit Tests Execution Above JDK 11

### PR DESCRIPTION
**Pull request by ParaskP7**
FYI: This PR resolves an issue similar to [this](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2540), by applying a fix similar to [this](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2559). PS: Thanks @wzieba ! ❤️ 🙇 🚀 

-----

As per the above, this issue happens when running unit tests above JDK 11 and, the reason for this failure is `ByteBuddy` from Mockito.

- Replacing `nhaarman` with `mockito-kotlin` fixed the issue.
- Bumping `mockito` to latest `4.8.1` is optional, but high recommended (our current version is `3.3.3`, a very outdated version).

This PR fixes failing unit tests execution on JDKs higher than 11 by replacing `nhaarman` with `mockito-kotlin` and updating `mockito` to latest stable. With this update `nhaarman` is effectively removed from the codebase in favor of `mockiot-kotlin`.

-----

Review instructions:

- This b5e383dfdd5289a711a6ae67bd6839d774fcc926 was mainly done automatically, so reviewing all the code and import changes is not absolutely necessary.
- Firstly, @oguzkocer and @wzieba I am adding you two as reviewers to this PR, and that is, because we were all involved in the `FluxC` suchlike PR too, so it should be pretty straightforward for us to test and merge this. In addition to that, I am adding @AjeshRPai here too in order for us to get a sign-off from someone from WPAndroid as well.

-----

To test:

- Firstly, make sure you can reproduce the issue. Install any JDK above 11 and update `JAVA_HOME` (if you use terminal) or settings in Android Studio (if you use AS). `Hint: Android Studio Eel has embedded JDK version 17`
- Checkout `trunk`.
- Run unit tests.
- Assert unit tests fail with something like: `Mockito cannot mock this class (...)`
- Checkout this branch: `test/fix-unit-tests-execution-above-jdk11`
- Run unit tests again.
- Assert unit tests pass.

-----

## Regression Notes
1. Potential unintended areas of impact

Test are not running on all possible combinations of available and used JDKs.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` section above.

3. What automated tests I added (or what prevented me from doing so)

N/A

-----

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -55,8 +55,8 @@ ext {
     androidxTestCoreVersion = '1.4.0'
     assertJVersion = '3.11.1'
     espressoVersion = '3.4.0'
-    mockitoCoreVersion = "3.3.3"
-    nhaarmanMockitoVersion = "2.2.0"
+    mockitoVersion = "4.8.1"
+    mockitoKotlinVersion = '4.0.0'
```
Comment (wzieba) [inline build.gradle:59]: ❓ Shouldn't we leave those versions to be declared by the version catalog?

**Review by wzieba — COMMENTED** [review]
The code looks good, but for me, some unit tests fail (89% successful). I can share the test report internally if it'll help.

I've run `testJetWasabiDebugUnitTest` task.

**Comment by ParaskP7** [comment]
👋 @wzieba !

Thank you for the review and testing, much appreciated! 🙇 💯 🥇 

> The code looks good, but for me, some unit tests fail (89% successful). I can share the test report internally if it'll help.
> 
> I've run testJetWasabiDebugUnitTest task.

Hmmm... that is indeed strange, can you try running `./gradlew testWordPressWasabiDebugUnitTest` instead? I am not sure why you chose to run the `testJetWasabiDebugUnitTest`, and I guess, you meant `testJetpackWasabiDebugUnitTest`, right? 🤔

Also, CI seems to be successful in running those tests as well... so I am 🤷 on what is happening on your side. Maybe we can pair-program on it on your side and I can take a look.

File: build.gradle (hunk 1 above)
Comment (ParaskP7) [reply build.gradle:59]: This is a nice idea @wzieba , but `version catalog` is only configured for the `FluxC` for the time being. Neither WPAndroid, nor WCAndroid are utilizing `version catalog` atm. Is that what you meant? 🤔

As such, for now, I wouldn't involve doing anything version catalog related.

**Comment by wzieba** [comment]
> I am not sure why you chose to run the testJetWasabiDebugUnitTest, and I guess, you meant testJetpackWasabiDebugUnitTest, right? 🤔

It's the same 😅 Gradle will resolve "testJet(...)" to "testJetpack(...)". Tests fail for me even on WordPress variant 😕 .

> Also, CI seems to be successful in running those tests as well... so I am 🤷 on what is happening on your side.

That's very strange. Just to make sure: with JDK above 11, those tests work for you on your local machine, do they? For me, they fail on completely fresh `git clone` 🤔 .

File: build.gradle (hunk 1 above)
Comment (wzieba) [reply build.gradle:59]: Oh sorry! I missed that "tiny" detail. Thanks 😅

**Comment by ParaskP7** [comment]
> It's the same 😅 Gradle will resolve "testJet(...)" to "testJetpack(...)". Tests fail for me even on WordPress variant 😕 .

Ah yes, true, I just never used `Jet` before and wasn't sure if that resolved to `Jetpack` automagically... But, yes, more importantly use the `WordPress` variant for testing, it might be better so we get consistent results, at least the two of us! 😅 

> That's very strange. Just to make sure: with JDK above 11, those tests work for you on your local machine, do they? For me, they fail on completely fresh git clone 🤔 .

Yes, they do. 👍

After me starting to use M1 with JDK 17, I noticed that those tests start failing for me too. Then, I saw your PR on the `FluxC` side and tried to apply the same for WPAndroid. After applying that fix and this branch of mine here, all those start passing again, give or take 3 tests that always fail anyway (see `WordPressPublicDataTest` and `JetpackPublicDataTest`), but those fail because of the local mismatch on the `package id` and `public hash key`, thus are unrelated to this change (they used to fail locally before well anyway).

**Comment by wzieba** [comment]
Gotcha, thanks for confirmation @ParaskP7 !

> After applying that fix and this branch of mine here, all those start passing again, give or take 3 tests that always fail anyway

Except of those two tests you mention, 411 more tests fail for me. I think most of them (maybe all?) with this message

```
org.mockito.exceptions.misusing.UnfinishedStubbingException: 
Unfinished stubbing detected here:
-> at org.wordpress.android.ui.reader.ReaderPostDetailUiStateBuilderTest.buildPostUiState(ReaderPostDetailUiStateBuilderTest.kt:419)

E.g. thenReturn() may be missing.
Examples of correct stubbing:
    when(mock.isOk()).thenReturn(true);
    when(mock.isOk()).thenThrow(exception);
    doThrow(exception).when(mock).someVoidMethod();
Hints:
 1. missing thenReturn()
 2. you are trying to stub a final method, which is not supported
 3. you are stubbing the behaviour of another mock inside before 'thenReturn' instruction is completed
```

Maybe it'll be useful for the future reference if someone will experience similar fails.

**Comment by ParaskP7** [comment]
> Yes, they do. 👍

FYI @wzieba , see below, I just tried running them again:

```
> Task :WordPress:testWordpressWasabiDebugUnitTest
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended

org.wordpress.android.sharedlogin.JetpackPublicDataTest > Should return correct release public hash key FAILED
    org.junit.ComparisonFailure at JetpackPublicDataTest.kt:14

org.wordpress.android.sharedlogin.JetpackPublicDataTest > Should return correct current package ID FAILED
    org.junit.ComparisonFailure at JetpackPublicDataTest.kt:21

org.wordpress.android.sharedlogin.WordPressPublicDataTest > Should return correct current package ID FAILED
    org.junit.ComparisonFailure at WordPressPublicDataTest.kt:14

3983 tests completed, 3 failed, 46 skipped

> Task :WordPress:testWordpressWasabiDebugUnitTest FAILED

FAILURE: Build failed with an exception.
```

**Comment by ParaskP7** [comment]
> Except of those two tests you mention, 411 more tests fail for me. I think most of them (maybe all?) with this message

Hmmm... this is indeed interesting, I am not sure why, but so many thanks for sharing! 🤔 🙇 👀

I also tried to run the `ReaderPostDetailUiStateBuilderTest` test suite on AS and everything passes as expected... 🤔 

> Maybe it'll be useful for the future reference if someone will experience similar fails.

Yes, definitely, and again, thanks for sharing! 🙏

**Comment by ParaskP7** [comment]
> ReaderPostDetailUiStateBuilderTest.kt:419

FYI @wzieba , this file is `350`/`351` lines long (see [here](https://github.com/wordpress-mobile/WordPress-Android/blob/619d38a51202f858e930063027d22b6cb3ba5c61/WordPress/src/test/java/org/wordpress/android/ui/reader/ReaderPostDetailUiStateBuilderTest.kt)), I am not sure why it is pointing to line `419` when it fails on your side... 🤔

**Comment by ParaskP7** [comment]
> ReaderPostDetailUiStateBuilderTest.kt:419

Btw @wzieba , and since it is only failing on your side (so far), is there anything you can do to fix this test suite? Have you tried applying the `thenReturn() may be missing` suggestion somehow? Have you tried running this test suite on AS, after updating AS to point to the same JDK (via `Build, Execution, Deployment > Build Tools > Gradle > Gradle JDK`? 🤔

Apologies for throwing all that towards you, but I am out of ideas on my side, especially since I can't reproduce it easily. Maybe one thing I could do is cloning the repo from scratch myself and try out this branch, but, I doubt this will produce a different result. No matter, I will indeed try that out at some point.

**Comment by wzieba** [comment]
> FYI @wzieba , this file is 350/351 lines long (see [here](https://github.com/wordpress-mobile/WordPress-Android/blob/619d38a51202f858e930063027d22b6cb3ba5c61/WordPress/src/test/java/org/wordpress/android/ui/reader/ReaderPostDetailUiStateBuilderTest.kt)), I am not sure why it is pointing to line 419 when it fails on your side... 🤔

That's a good eye! I don't know what happens there 😅 .

> Apologies for throwing all that towards you, but I am out of ideas on my side, especially since I can't reproduce it easily.

In the meantime, I've created a sample CI environment for reproduction, and those tests fail on this machine too :( look [here](https://github.com/wzieba/WordPressTestsCheck/actions/runs/3469711739/jobs/5797032052)

Maybe let's continue internally on Slack?

**Comment by ParaskP7** [comment]
> In the meantime, I've created a sample CI environment for reproduction, and those tests fail on this machine too :( look [here](https://github.com/wzieba/WordPressTestsCheck/actions/runs/3469711739/jobs/5797032052)

Awesome, thank you for doing that, I am now seeing the test failures you are mentioning more clearly. 💯 

Btw, looking at your GitHub Action [configuration](https://github.com/wzieba/WordPressTestsCheck/blob/main/.github/workflows/main.yml), I am seeing you are using [java-version = '18'](https://github.com/wzieba/WordPressTestsCheck/blob/main/.github/workflows/main.yml#L14). @wzieba , can we try with `17` instead? 🤔 

> Maybe let's continue internally on Slack?

Sure thing, I guess that's enough of PR spamming already! 😅

**Review by wzieba — APPROVED** [review]
As discussed internally p1668510432514089-slack-C030U03RC8Y:

- this PR fixes Mockito issue for JDK 17
- this PR does not fix Mockito issue for JDK 18 but we do not worry as JDK 18 is not LTS
- we don't know about JDK 19 (next LTS) a even Gradle version currently used doesn't support JDK 19.

Thanks @ParaskP7 for debugging together. This PR looks good to me 👍 !

**Comment by ParaskP7** [comment]
> Thanks @ParaskP7 for debugging together. This PR looks good to me 👍 !

Awesome, thank you so much for the great summary and for debugging this @wzieba , you are an awesome GitHub Actioner, and you know that, kudos for setting the testing environment and all, much appreciated! 🥇 🥇 🥇

**Review by oguzkocer — APPROVED** [review]
Using JDK17, I verified the failures in `trunk` and the fix in this branch.

For full disclosure; I had one failure, but it's unrelated to this PR.

```
org.wordpress.android.sharedlogin.JetpackPublicDataTest > Should return correct release public hash key FAILED
    org.junit.ComparisonFailure at JetpackPublicDataTest.kt:14
```

**Comment by ParaskP7** [comment]
👋 @oguzkocer !

Thank you so much for the extra review and testing Oguz, much appreciated the extra 👍 here! 🙇 🥇 💯 

> For full disclosure; I had one failure, but it's unrelated to this PR.

Yea, the failures on both, the `WordPressPublicDataTest` and `JetpackPublicDataTest` test suites, is something to be expected, see my comment from [above](https://github.com/wordpress-mobile/WordPress-Android/pull/17459#issuecomment-1315111000), and quoting the specifics of it below.

> ...give or take 3 tests that always fail anyway (see `WordPressPublicDataTest` and `JetpackPublicDataTest`), but those fail because of the local mismatch on the package id and public hash key, thus are unrelated to this change (they used to fail locally before well anyway).

-----

@RenanLukas I wonder if we could do something about them? 🤔

I mean, these test failing, mind the fact, rightfully so, still, they are giving us a false negative, which is confusing for engineers that haven't seen that before. I know I keep ignoring them when running the `testWordpressWasabiDebugUnitTest` task since I know this uses another `package id` and `hash key` anyway, still, this is something that I wished I shouldn't had to care about and see all tests passing instead.

Also, I am now seeing that we would be needing to explain why this test are failing to others, and educate people to not worry about them too much, especially in cases like this. Maybe this is something we can improve on, with some kind of automation, additional documentation, the failing test explaining the reason more clearly, or otherwise, wdyt? 🤔

**Pull request merged**